fuel-vm/src/constraints/reg_key.rs

```diff
@@ -0,0 +1,342 @@
+    SP, sp
+    FP, fp
+    HP, hp
+    ERR, err
+    GGAS, ggas
+    CGAS, cgas
+    BAL, bal
+    IS, is
+    RET, ret
+    RETL, retl
+    FLAG, flag
+}
+
+pub(crate) struct ReadRegisters<'a> {
+    pub(crate) zero: RegMut<'a, ZERO>,
+    pub(crate) one: RegMut<'a, ONE>,
+    pub(crate) of: RegMut<'a, OF>,
+    pub(crate) pc: RegMut<'a, PC>,
+    pub(crate) ssp: RegMut<'a, SSP>,
+    pub(crate) sp: RegMut<'a, SP>,
+    pub(crate) fp: RegMut<'a, FP>,
+    pub(crate) hp: RegMut<'a, HP>,
+    pub(crate) err: RegMut<'a, ERR>,
+    pub(crate) ggas: RegMut<'a, GGAS>,
+    pub(crate) cgas: RegMut<'a, CGAS>,
+    pub(crate) bal: RegMut<'a, BAL>,
+    pub(crate) is: RegMut<'a, IS>,
+    pub(crate) ret: RegMut<'a, RET>,
+    pub(crate) retl: RegMut<'a, RETL>,
+    pub(crate) flag: RegMut<'a, FLAG>,
+}
+
+pub(crate) struct ReadRegistersRef<'a> {
+    pub(crate) zero: Reg<'a, ZERO>,
+    pub(crate) one: Reg<'a, ONE>,
+    pub(crate) of: Reg<'a, OF>,
+    pub(crate) pc: Reg<'a, PC>,
+    pub(crate) ssp: Reg<'a, SSP>,
+    pub(crate) sp: Reg<'a, SP>,
+    pub(crate) fp: Reg<'a, FP>,
+    pub(crate) hp: Reg<'a, HP>,
+    pub(crate) err: Reg<'a, ERR>,
+    pub(crate) ggas: Reg<'a, GGAS>,
+    pub(crate) cgas: Reg<'a, CGAS>,
+    pub(crate) bal: Reg<'a, BAL>,
+    pub(crate) is: Reg<'a, IS>,
+    pub(crate) ret: Reg<'a, RET>,
+    pub(crate) retl: Reg<'a, RETL>,
+    pub(crate) flag: Reg<'a, FLAG>,
+}
+pub(crate) struct WriteRegisters<'a>(pub &'a mut [Word; VM_REGISTER_WRITE_COUNT]);
+pub(crate) struct WriteRegistersRef<'a>(pub &'a [Word; VM_REGISTER_WRITE_COUNT]);
+
+pub(crate) fn split_registers(registers: &mut [Word; VM_REGISTER_COUNT]) -> (ReadRegisters<'_>, WriteRegisters<'_>) {
+    let [zero, one, of, pc, ssp, sp, fp, hp, err, ggas, cgas, bal, is, ret, retl, flag, rest @ ..] = registers;
+    let r = ReadRegisters {
+        zero: RegMut(zero),
+        one: RegMut(one),
+        of: RegMut(of),
+        pc: RegMut(pc),
+        ssp: RegMut(ssp),
+        sp: RegMut(sp),
+        fp: RegMut(fp),
+        hp: RegMut(hp),
+        err: RegMut(err),
+        ggas: RegMut(ggas),
+        cgas: RegMut(cgas),
+        bal: RegMut(bal),
+        is: RegMut(is),
+        ret: RegMut(ret),
+        retl: RegMut(retl),
+        flag: RegMut(flag),
+    };
+    (r, WriteRegisters(rest))
+}
+
+pub(crate) fn copy_registers(
+    read_registers: &ReadRegistersRef<'_>,
+    write_registers: &WriteRegistersRef<'_>,
+) -> [Word; VM_REGISTER_COUNT] {
+    [
+        *read_registers.zero,
+        *read_registers.one,
+        *read_registers.of,
+        *read_registers.pc,
+        *read_registers.ssp,
+        *read_registers.sp,
+        *read_registers.fp,
+        *read_registers.hp,
+        *read_registers.err,
+        *read_registers.ggas,
+        *read_registers.cgas,
+        *read_registers.bal,
+        *read_registers.is,
+        *read_registers.ret,
+        *read_registers.retl,
+        *read_registers.flag,
+        write_registers.0[0],
```

Comment:
You can copy slices or use [concat-arrays](https://github.com/canndrew/concat-arrays). The compiler should be clever enough to optimize it

Reply:
That would require initializing a blank array to begin with.

Reply:
The compiler can usually optimize the blank initialization though if it knows it will later be fully allocated. @xgreenx and I verified this behavior when he implemented the `double_key!` type.

Reply:
Also, hardcoding the register count like this seems error-prone, and it would be nice to avoid that